framework/Util/Cron/TTimeScheduler.php

```diff
@@ -0,0 +1,856 @@
+		foreach (explode(',', $matches[6]) as $match) {
+			if (preg_match('/^(\*|\?|' . $dow . ')(L)?(?:\-(' . $dow . ')(L)?)?(?:\/(' . $dow . '))?(?:#([1-5]))?$/i', $match, $m2)) {
+				if ($m2[1] === '*' || $m2[1] === '?') {
+					$data = ['dow' => 0, 'end' => 6];
+				} else {
+					$data = ['dow' => (int) $this->translateDayOfWeek($m2[1])];
+					$data['end'] = (isset($m2[3]) && $m2[3]) ? (int) $this->translateDayOfWeek($m2[3]) : $data['dow'];
+				}
+				$data['lastDow'] = $m2[2] ?? null;
+				$data['lastEnd'] = $m2[4] ?? null;
+				$data['period'] = (int) max($m2[5] ?? 1, 1);
+				$data['week'] = $m2[6] ?? null;
+				if (!($m2[3] ?? 0) && ($m2[5] ?? 0)) {
+					$data['end'] = 6; //No end with period
+				}
+				$this->_attr[self::DAY_OF_WEEK][] = $data;
+			} else {
+				throw new TInvalidDataValueException('timescheduler_invalid_string', $match);
+			}
+		}
+			
+		$this->_attr[self::YEAR] = [];
+		$matches[7] = $matches[7] ?? '*';
+		foreach (explode(',', $matches[7]) as $match) {
+			if (preg_match('/^(\*|' . $year . ')(?:\-(' . $year . '))?(?:\/([1-9]?[0-9]))?$/i', $match, $m2)) {
+				if ($m2[1] === '*') {
+					$data = ['year' => self::YEAR_MIN];
+					$data['end'] = self::YEAR_MAX;
+				} else {
+					$data = ['year' => $m2[1]];
+					$data['end'] = $m2[2] ?? $m2[1];
+				}
+				$data['period'] = max($m2[3] ?? 1, 1);
+				if (!($m2[2] ?? 0) && ($m2[3] ?? 0)) {
+					$data['end'] = self::YEAR_MAX; //No end with period
+				}
+				$this->_attr[self::YEAR][] = $data;
+			} else {
+				throw new TInvalidDataValueException('timescheduler_invalid_string', $match);
+			}
+		}
+	}
+	
+	protected function translateMonthOfYear($v)
+	{
+		if (is_numeric($v)) {
+			return $v;
+		}
+		foreach (self::$_keywords[self::MONTH_OF_YEAR] as $moy => $regex) {
+			if (preg_match('/^(' . $regex . ')$/i', $v)) {
+				return $moy;
+			}
+		}
+		return $v;
+	}
+	
+	protected function translateDayOfWeek($v)
+	{
+		if (is_numeric($v)) {
+			return $v;
+		}
+		foreach (self::$_keywords[self::DAY_OF_WEEK] as $dow => $regex) {
+			if (preg_match('/^(' . $regex . ')$/i', $v)) {
+				return $dow;
+			}
+		}
+		return $v;
+	}
+	
+	
+	/**
+	 * Given which minutes are valid for the parsed time stamp, this generates
+	 * an array of valid minutes.  Each minute is available as the key to the array
+	 * and the value contains true or false whether the minute meets the cron
+	 * time stamp
+	 */
+	protected function getMinutesArray()
+	{
+		$ma = array_pad([], 60, null);
+		foreach ($this->_attr[self::MINUTE] as $m) {
+			if (is_numeric($m['min'])) {
+				for ($i = $m['min']; $i <= $m['end'] && $i < 60; $i += $m['period']) {
+					$ma[$i] = 1;
+				}
+			}
+		}
+		return $ma;
+	}
+	
+	
+	/**
+	 * Given which hours are valid for the parsed time stamp, this generates
+	 * an array of valid hours.  Each hour is available as the key to the array
+	 * and the value contains true or false whether the hour meets the cron
+	 * time stamp
+	 */
+	protected function geHoursArray()
```

Comment:
This should probably be named getHoursArray(), with and added "t"

Reply:
ah yeas.  true

Reply:
good eye on this.